oak/server/time/roughtime_client.cc

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright 2019 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "roughtime_client.h"
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <inttypes.h>
+#include <netdb.h>
+#include <netinet/udp.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include <array>
+#include <string>
+#include <vector>
+
+#include "absl/memory/memory.h"
+#include "absl/strings/escaping.h"
+#include "absl/strings/str_format.h"
+#include "asylo/util/logging.h"
+#include "asylo/util/status.h"
+#include "asylo/util/status_macros.h"
+#include "client.h"
+#include "oak/common/nonce_generator.h"
+
+using ::asylo::Status;
+using ::asylo::StatusOr;
+
+namespace oak {
+
+// Information we need about roughtime servers.
+// Assume we will only support Ed25519 public key and UDP for now.
+struct RoughtimeServerSpec {
+  std::string name;
+  std::string host;
+  std::string port;
+  std::string public_key_base64;
+};
+
+struct RoughtimeInterval {
+  roughtime::rough_time_t min;
+  roughtime::rough_time_t max;
+};
+
+const std::vector<RoughtimeServerSpec> servers{{"Google", "roughtime.sandbox.google.com", "2002",
+                                                "etPaaIxcBMY1oUeGpwvPMCJMwlRVNxv51KK/tktoJTQ="},
+                                               {"Cloudflare", "roughtime.cloudflare.com", "2002",
+                                                "gD63hSj3ScS+wuOeGrubXlq35N1c5Lby/S+T7MNTjxo="}};
+
+// The minimum number of overlapping intervals we need to trust the time.
+constexpr int kMinOverlappingTimeIntervals = 2;
+
+// Number of seconds that we will wait for a reply from the server.
+constexpr int kTimeoutSeconds = 3;
+
+// Number of times we will retry connecting to the server.
+constexpr int kServerRetries = 3;
+
+// The size of the receive buffer.
+// This seems to be the same as the minimum request size according to the Roughtime samples.
+// eg. https://roughtime.googlesource.com/roughtime/+/refs/heads/master/simple_client.cc#250
+constexpr size_t kReceiveBufferSize = roughtime::kMinRequestSize;
+
+// Maximum radius accepted for a roughtime response (1 minute in microseconds).
+constexpr uint32_t kMaxRadius = 60000000;
```

Comment:
Rename to something that makes it clear what the unit is (e.g. `*Microseconds`)

Reply:
Done